dds/DCPS/MemoryPool.cpp

```diff
@@ -0,0 +1,699 @@
+    set_larger_free(NULL, NULL);
+  }
+}
+FreeHeader*
+FreeHeader::smaller_free(unsigned char* pool_base) const
+{
+  FreeHeader* result = NULL;
+  if (offset_smaller_free_ != ULONG_MAX) {
+    result = reinterpret_cast<FreeHeader*>(pool_base + offset_smaller_free_);
+  }
+  return result;
+}
+
+FreeHeader*
+FreeHeader::larger_free(unsigned char* pool_base) const
+{
+  FreeHeader* result = NULL;
+  if (offset_larger_free_ != ULONG_MAX) {
+    result = reinterpret_cast<FreeHeader*>(pool_base + offset_larger_free_);
+  }
+  return result;
+}
+
+void
+FreeHeader::set_smaller_free(FreeHeader* next, unsigned char* pool_base)
+{
+  if (next) {
+    offset_smaller_free_ = reinterpret_cast<unsigned char*>(next) - pool_base;
+  } else {
+    offset_smaller_free_ = ULONG_MAX;
+  }
+}
+
+void
+FreeHeader::set_larger_free(FreeHeader* prev, unsigned char* pool_base)
+{
+  if (prev) {
+    offset_larger_free_ = reinterpret_cast<unsigned char*>(prev) - pool_base;
+  } else {
+    offset_larger_free_ = ULONG_MAX;
+  }
+}
+
+FreeIndexNode::FreeIndexNode()
+: size_(0)
+, limit_(0)
+, ptr_(0)
+{
+}
+
+void
+FreeIndexNode::set_sizes(size_t size, size_t limit)
+{
+  size_ = size;
+  limit_ = limit;
+}
+
+FreeIndex::FreeIndex(FreeHeader*& largest_free)
+: size_(0)
+, largest_free_(largest_free)
+{
+}
+
+void
+FreeIndex::add(FreeHeader* freed)
+{
+  FreeIndexNode* node = nodes_;
+  do {
+    // If the freed block is of the size this node manages
+    if (node->contains(freed->size())) {
+      if ((node->ptr() == NULL) || (node->ptr()->size() >= freed->size())) {
+        node->set_ptr(freed);
+      }
+      break;
+    }
+    ++node;
+  } while (node < nodes_ + size_);
+}
+
+void
+FreeIndex::remove(FreeHeader* free_block, FreeHeader* larger)
+{
+  FreeIndexNode* node = nodes_;
+  do {
+    if (node->contains(free_block->size())) {
+      if (node->ptr() == free_block) {
+        if (larger && node->contains(larger->size())) {
+          node->set_ptr(larger);
+        } else {
+          node->set_ptr(NULL);
+        }
+      }
+      break;
+    }
+    ++node;
+  } while (node < nodes_ + size_);
+}
+
+void
+FreeIndex::init(FreeHeader* init_free_block)
+{
+  int init_index = 0;
+  unsigned int size = 8;
+  while (size <= 4096) {
+    nodes_[size_].set_sizes(size, (size == 4096) ? -1 :  size*2);
+    if (init_free_block->size() >= size) {
+      init_index = size_;
+    }
+    ++size_;
+    size *= 2;
+  }
+  nodes_[init_index].set_ptr(init_free_block);
+}
+
+FreeHeader*
+FreeIndex::find(size_t size, unsigned char* pool_base)
+{
+  // Search index
+  FreeIndexNode* index = nodes_ + size_ - 1;
+
+  // Larger or equal to size
+  FreeHeader* result = NULL;
+  if (largest_free_ && (largest_free_->size() >= size)) {
+    result = largest_free_;
+
+    // Look from largest to smallest
+    while (index >= nodes_) {
+      if (index->ptr() && index->ptr()->size() >= size) {
+        result = index->ptr();
+      }
+
+      if (index->size() < size) {
```

Comment:
actually, the code is correct.  The check is after the node is used, and before the node is decremented.